src/agent/src/image.rs

```diff
@@ -0,0 +1,318 @@
+            Ok(image) => {
+                info!(
+                    sl(),
+                    "pull and unpack image {image:?}, cid: {cid:?} succeeded."
+                );
+            }
+            Err(e) => {
+                error!(
+                    sl(),
+                    "pull and unpack image {image:?}, cid: {cid:?} failed with {:?}.",
+                    e.to_string()
+                );
+                return Err(e);
+            }
+        };
+        self.add_image(String::from(image), String::from(cid)).await;
+        Ok(format! {"{}/rootfs",bundle_path.display()})
+    }
+
+    /// When being passed an image name through a container annotation, merge its
+    /// corresponding bundle OCI specification into the passed container creation one.
+    pub async fn merge_bundle_oci(&self, container_oci: &mut oci::Spec) -> Result<()> {
+        if let Some(image_name) = container_oci.annotations.get(ANNO_K8S_IMAGE_NAME) {
+            let images = self.images.lock().await;
+            if let Some(container_id) = images.get(image_name) {
+                let image_oci_config_path = Path::new(CONTAINER_BASE)
+                    .join(container_id)
+                    .join(CONFIG_JSON);
+                debug!(
+                    sl(),
+                    "Image bundle config path: {:?}", image_oci_config_path
+                );
+
+                let image_oci =
+                    oci::Spec::load(image_oci_config_path.to_str().ok_or_else(|| {
+                        anyhow!(
+                            "Invalid container image OCI config path {:?}",
+                            image_oci_config_path
+                        )
+                    })?)
+                    .context("load image bundle")?;
+
+                if let (Some(container_root), Some(image_root)) =
+                    (container_oci.root.as_mut(), image_oci.root.as_ref())
+                {
+                    let root_path = Path::new(CONTAINER_BASE)
+                        .join(container_id)
+                        .join(image_root.path.clone());
+                    container_root.path = String::from(root_path.to_str().ok_or_else(|| {
+                        anyhow!("Invalid container image root path {:?}", root_path)
+                    })?);
+                }
+
+                if let (Some(container_process), Some(image_process)) =
+                    (container_oci.process.as_mut(), image_oci.process.as_ref())
+                {
+                    self.merge_oci_process(container_process, image_process);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Partially merge an OCI process specification into another one.
+    fn merge_oci_process(&self, target: &mut oci::Process, source: &oci::Process) {
+        if target.args.is_empty() && !source.args.is_empty() {
+            target.args.append(&mut source.args.clone());
+        }
+
+        if target.cwd == "/" && source.cwd != "/" {
```

Comment:
I think all the paths should be resolved as we should not trust inputs.

Reply:
Fixed.

Reply:
Same here, @ChengyuZhu6 . Could you please push the branch with the fix?

Reply:
@wainersm @jodh-intel I fixed it ago. A potential issue with the safe-path crate is that its `scoped_join(root, unsafe_path)` function requires root to be an existing directory. This implies that we have to create the directory in this part of the code, which is not desirable from a logical perspective.